### DISCUSSION TITLE
Replacing "cmd_" calls with decorators

### DISCUSSION THREAD

**Discussion by elParaguayo** [qtile/qtile#2774] [Show and tell]
This is in relation to PR #2727 

# Using decorators to replace 'cmd_' syntax

So, I've been working on the decorators and have a working system with 100% of tests passing.

During the refactoring of the code, I've had to make some adjustments which I've set out here. I would welcome people's thoughts.

Deep breath, there's a lot here...

### CommandObject

Main change is to add a `@CommandObject.expose` decorator. This is the decorator that must be added to methods that are to be exposed via IPC.

As this doesn't work by parsing the method name, subclasses need to add the decorator even if the parent had decorated the method. i.e. if a widget implements and `info` method, that will need the decorator.

I've also added a second decorator, largely for convenience purposes. `@CommandObject.map_command("name")` allows you to bind another command called `name` to the decorated method. This was helpful when doing layouts where you had lines like this in layouts:
``` python
cmd_previous = _SimpleLayoutBase.previous
cmd_next = _SimpleLayoutBase.next

cmd_up = cmd_previous
cmd_down = cmd_next
```
We can't rewrite this as:
``` python
next = _SimpleLayoutBase.next
```
because `next` is reserved python function. But you can do:
``` python
@CommandObject.expose
@CommandObject.map_command("previous")
def up(self):
    _SimpleLayoutBase.previous(self)

@CommandObject.expose
@CommandObject.map_command("next")
def down(self):
    _SimpleLayoutBase.next(self)
```
Obviously, we could write this as four methods, rather than two. I don't have strong feelings about it.

### Deprecations

I've added some code in `CommandObject.__getattr_` to handle calls to `cmd_` methods. The methods don't exist but, instead, I look up whether the command name exists without `cmd_` and, if it does, return it. The user will get a warning too (I tried using `warnings.warn...` but couldn't see this in my logs, so used `logger.warning` instead.)

In a similar vein, if a command is called `cmd_`, the code will allow it but the user will get a message telling them to use the decorator instead.

### interface.py

There's also a bit of care needed in `libqtile/command/interface.py` when we retrieve the method. Depending on the call, this may be bound already (this happens when the slice layout tries to retrieve commands for the fallback layout (more on this below)), or unbound. So, I handle both cases currently:
``` python
if hasattr(cmd, "__self__"):
    return SUCCESS, cmd(*args, **kwargs)
else:
    # If not, pass object as first argument
    return SUCCESS, cmd(obj, *args, **kwargs)
```

## Backend

There are three problematic commands on X11: `cmd_opacity`, `cmd_hints` and `cmd_match`. All three make calls to properties/methods without the `cmd_" prefix.

I haven't changed any of these yet.

For `cmd_opacity` I'd be inclined to change the command name to `set_opacity` and then do a migrate command if we need.

For the other two, I'm not so sure but I suspect changing the command name is simpler than changing the underlying calls. I don't know how much these calls are actually used.

## Manager/Core

`cmd_groups` returns the keys from the `self.groups` dict. 
`cmd_screens` provides more information on each screen in `self.screens`.

These are unchanged for now and tests aren't failing (which probably means I need to add a test!)

I'd be inclined to change the command name and retain the properties.

### Matrix

`cmd_add()` does not do the same thing as `add()`. The former adds a column whereas the latter adds a client.  I renamed command to `add_column` and updated test. This will need a migrate script.

Question: should `delete` be renamed to `delete_column` for consistency? There's no clash currently.

### Slice

This one was painful!

Main issue was that the layout needs to find commands for the `fallback` layout. Previously this was handled via `__getattr__`:
``` python
def __getattr__(self, name):
    """Delegate unimplemented command calls to active layout.
    For ``cmd_``-methods that don't exist on the Slice class, this looks
    for an implementation on the active layout.
    """
    if name.startswith('cmd_'):
        return getattr(self._get_active_layout(), name)
    return super().__getattr__(name)
```
I've removed that method and replaced it with this:
``` python
def command(self, name: str):
    if name in self._commands:
        return self._commands.get(name)

    # We're returning a bound method here
    return getattr(self._get_active_layout(), name)
```
Basically, I check if the command belongs to the main slice layout and, if not, I see if it's in the fallback layout.

### Stack

Same `cmd_add` issue. This has been renamed to `add_stack`.

Problem here is if someone had a keybinding for `lazy.layout.add()`, this would have worked on Matrix and Stack but won't now.

### Tile

This one was silly.

`cmd_shuffle_up` calls `self.up` but `cmd_up` calls `previous`. Same for `down`. 

Tidied this up.

### Xmonad

`cmd_grow` calls `self.grow` (similar for `cmd_shrink`). The `self.` methods are called multiple times so I renamed these `self._grow` and `self._shrink`.

### Notify

Similar to Xmonad, widget has `cmd_clear` calling `self.clear` but the method is used elsewhere. Renamed to `self._clear` to prevent exposure of ability to pass arguments (original `clear` method takes an optional `reason` argument).

### Stock ticker

One config argument is called `function`. This clashes with the old `cmd_function` that is inherited by all `CommandObject`s. Renamed argument so will need a migrate script.

## Tests

I need to add a test to check whether deprecated calls still work and will add an explicit test for the decorators.

Let me know if there's anything else you think should be tested.

### Documentation

Haven't started updating yet...

**Comment by m-col** [comment]
> @CommandObject.map_command("next")

Nice solution you've made here. Perhaps `@CommandObject.expose_as("next")` for consistency?

> There are three problematic commands on X11: cmd_opacity, cmd_hints and cmd_match. All three make calls to properties/methods without the `cmd_" prefix.

As the lazy commands are the 'canonical' API I'd prefer if we avoid touching those if we can, instead modifying the properties. Those will only be used by people more comfortable at Python and hacking into Qtile, so they'll be more likely to be aware of migrations and of debugging a sudden config breakage. So perhaps we can just prefix the properties with underscores or something.

> cmd_groups returns the keys from the self.groups dict. cmd_screens provides more information on each screen in self.screens.

Same argument as above really, though I guess these seem a bit more tricky. People could be using these commands in custom scripts over IPC, and I'm not sure if migrations would work then. However, these same people must be comfortable enough with Python so might be alright as debugging things, and nobody is going to have `lazy.groups()` bound to a keybinding so changing those cmds might not be too distruptive.

> Layouts

Similarly as above, I think it's high priority to not break things, even if that means having not-so-clean internal code. I don't think peope will be calling a layout's non-cmd methods because that may well break things(?) so I think the ideal scenario would be leaving all of the commands on all layouts untouched in their command name.

> Stock ticker

I think we can do better than a migration on this one:
```diff
diff --git a/libqtile/widget/stock_ticker.py b/libqtile/widget/stock_ticker.py
index 6c372978..a3863ac7 100644
--- a/libqtile/widget/stock_ticker.py
+++ b/libqtile/widget/stock_ticker.py
@@ -44,17 +44,23 @@ class StockTicker(GenPollUrl):
 
     defaults = [
         ("interval", "1min", "The default latency to query"),
-        ("function", "TIME_SERIES_INTRADAY", "The default API function to query"),
+        ("func", "TIME_SERIES_INTRADAY", "The default API function to query"),
+        ("function", "TIME_SERIES_INTRADAY", "DEPRECATED: Use `func`."),
     ]
 
     def __init__(self, **config):
+        if "function" in config:
+            logger.warning(
+                "`function` parameter is deprecated. Please rename to `func`"
+            )
+            config["func"] = config.pop("function")
         GenPollUrl.__init__(self, **config)
         self.add_defaults(StockTicker.defaults)
         self.sign = locale.localeconv()['currency_symbol']
         self.query = {
             "interval": self.interval,
             "outputsize": "compact",
-            "function": self.function
+            "function": self.func
         }
         for k, v in config.items():
             self.query[k] = v
```

Everything else looks really good, and thanks for the nice improvements. In general migrations should be a last resort, because we don't really know how many people are aware of that functionality (and they are still quite new), so any opportunity we have to not break things (even with the nice goal of cleaner code) we should take.

**Reply by elParaguayo** [reply]
Thanks. Some good suggestions there. I'll make some changes and update.

**Reply by elParaguayo** [reply]
Have addressed all of these and pushed updated. Looks like some conflicts now but I'll fix them later.

**Comment by tych0** [comment]
> There are three problematic commands on X11: cmd_opacity, cmd_hints and cmd_match. All three make calls to properties/methods without the `cmd_" prefix.

Can you expand a bit about why this matters? I'm not sure I understand.

**Reply by elParaguayo** [reply]
I think it's ok to ignore this comment as is addressed in latest updates.

The problem was that the command name clashed with an underlying attribute/method.
